rust/lance-core/src/datatypes/field.rs

```diff
@@ -376,15 +376,45 @@ impl Field {
             .collect::<Vec<_>>();
         if !children.is_empty() || filter(self) {
             Some(Self {
-                name: self.name.clone(),
-                id: self.id,
-                parent_id: self.parent_id,
-                logical_type: self.logical_type.clone(),
-                metadata: self.metadata.clone(),
-                encoding: self.encoding.clone(),
-                nullable: self.nullable,
                 children,
-                dictionary: self.dictionary.clone(),
+                ..self.clone()
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Create a new field by selecting fields by their ids.
+    ///
+    /// If a field has it's id in the list of ids then it will be included
+    /// in the new field. If a field is selected, all of it's parents will be
+    /// and all of it's children will be included.
+    ///
+    /// For example, for the schema:
+    ///
+    /// ```text
+    /// 0: x struct {
+    ///     1: y int32
+    ///     2: l list {
+    ///         3: z int32
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// If the ids are `[2]`, then this will include the parent `0` and the
+    /// child `3`.
+    pub(crate) fn project_by_ids(&self, ids: &[i32]) -> Option<Self> {
```

Comment:
This is a breaking change in the behavior of `project_by_ids`. Before, if you wanted to select a field and all of it's children, you had to pass the ids of all it's children. Now you can just pass the id of it's parent. it's still possible to select specific children by passing just the ids of the children you care about, and not the id of the parent.

This does make it impossible to select a parent field without any of it's children, but I don't think that's really sensible anyways. In fact, that's why I made this change: when trying to select a list field it would take that field without it's child field, and that would lead to an error later because a list always needs exactly one child.

This passes all other unit tests as-is, so I think this should be fine.